# Filter and clean data

When working with a dataset, oftentimes you're only interested in a smaller subset of your data. For example, say I have a car loan dataset and I want to filter out the data to only have a car type of Toyota Sienna with an interest rate of 7.02%. So the first thing I'm going to do is I'm look at the first five rows in my dataset, and while it appears that the first five rows are only Toyota Siennas, that doesn't mean the rest of my dataset is all of car type Toyota Sienna. 


### car_type filter
Comparison Operator | Meaning
--- | --- 
<  | less than
<= | less than or equal to
'>'| greater than
'>=' | greater than or equal to
== | equal
!= | not equal

In [2]:
# Import libraries
import pandas as pd
import numpy as np

# Load Excel File
filename = 'car_financing.xlsx'
df = pd.read_excel(filename)

In [4]:
df.head()

Month  Starting Balance  Repayment  Interest Paid  Principal Paid  \
0      1          34689.96     687.23         202.93          484.30   
1      2          34205.66     687.23         200.10          487.13   
2      3          33718.53     687.23         197.25          489.98   
3      4          33228.55     687.23         194.38          492.85   
4      5          32735.70     687.23         191.50          495.73   

   New Balance  term  interest_rate       car_type  
0     34205.66    60         0.0702  Toyota Sienna  
1     33718.53    60         0.0702  Toyota Sienna  
2     33228.55    60         0.0702  Toyota Sienna  
3     32735.70    60         0.0702  Toyota Sienna  
4     32239.97    60         0.0702  Toyota Sienna

The first thing I'm going to do is I'll use the value counts method on the car type column to see what other kind of cars I have my dataset. I have my data frame, I have square brackets, I have the column I'm interested in. I'm going to close those brackets, and then I have the value counts method. When I press shift enter, you'll see that I have 120 Toyota Siennas. 

In [5]:
# Let's first start by looking at the car_type column. 
df['car_type'].value_counts()

car_type
VW Golf R         144
Toyota Sienna     120
Toyota Carolla    111
Toyota Corolla     33
Name: count, dtype: int64

Say for example, I was interested in Toyota Corollas instead, I would have to fix the data entry errors because I have 111 Toyota Carollas, instead of Corollas. This is a really important thing to take note of, as oftentimes you'll have misspellings in your dataset, you'll have errors, you'll have things you don't quite understand, but that's all part of the data exploration process. 

And what I'll do now is I'll create a car filter, and the way this works is I have a data frame, I have square brackets, I have the column that I'm interested in. I close those single brackets. I have two equal signs because this is equality, and I have the type of car I'm interested in, in this case, Toyota Sienna. And what this produces is a pandas series of true and false values. 

In [6]:
# Notice that the filter produces a pandas series of True and False values
car_filter = df['car_type']=='Toyota Sienna'

In [8]:
car_filter.head()

0    True
1    True
2    True
3    True
4    True
Name: car_type, dtype: bool

There's a couple different ways to utilize this pandas filter of true and false values to get a data frame of just Toyota Siennas. One way is to have your data frame, square brackets, your car filter, which is your pandas series, you close those brackets, and then I'm just going to look at the first five rows.

In [12]:
# Approach 1 using square brackets
# Filter dataframe to get a DataFrame of only 'Toyota Sienna'
df[car_filter].head()

Month  Starting Balance  Repayment  Interest Paid  Principal Paid  \
0      1          34689.96     687.23         202.93          484.30   
1      2          34205.66     687.23         200.10          487.13   
2      3          33718.53     687.23         197.25          489.98   
3      4          33228.55     687.23         194.38          492.85   
4      5          32735.70     687.23         191.50          495.73   

   New Balance  term  interest_rate       car_type  
0     34205.66    60         0.0702  Toyota Sienna  
1     33718.53    60         0.0702  Toyota Sienna  
2     33228.55    60         0.0702  Toyota Sienna  
3     32735.70    60         0.0702  Toyota Sienna  
4     32239.97    60         0.0702  Toyota Sienna

The second way is to use a lock attribute, and the way this works is I have a data frame, I have .loc, I have single brackets, I have my pandas series of true and false values, and this colon just means that I want to look at all the columns and I'll press shift plus enter, and these two approaches are equivalent in the result they produce, but oftentimes the second approach is more legible. 

In [13]:
# Approach 2 using loc
# Filter dataframe to get a DataFrame of only 'Toyota Sienna'
df.loc[car_filter, :]

Month  Starting Balance  Repayment  Interest Paid  Principal Paid  \
0        1          34689.96     687.23         202.93          484.30   
1        2          34205.66     687.23         200.10          487.13   
2        3          33718.53     687.23         197.25          489.98   
3        4          33228.55     687.23         194.38          492.85   
4        5          32735.70     687.23         191.50          495.73   
..     ...               ...        ...            ...             ...   
115     56           3133.83     632.47           9.37          623.10   
116     57           2510.73     632.47           7.51          624.96   
117     58           1885.77     632.47           5.64          626.83   
118     59           1258.94     632.47           3.76          628.71   
119     60            630.23     632.47           1.88          630.59   

     New Balance  term  interest_rate       car_type  
0       34205.66    60         0.0702  Toyota Sienna  
1       33718.53    60         0.0702  Toyota Sienna  
2       33228.55    60         0.0702  Toyota Sienna  
3       32735.70    60         0.0702  Toyota Sienna  
4       32239.97    60         0.0702  Toyota Sienna  
..           ...   ...            ...            ...  
115      2510.73    60         0.0359  Toyota Sienna  
116      1885.77    60         0.0359  Toyota Sienna  
117      1258.94    60         0.0359  Toyota Sienna  
118       630.23    60         0.0359  Toyota Sienna  
119        -0.36    60         0.0359  Toyota Sienna  

[120 rows x 9 columns]

As you can see, I have identical outputs for the two different approaches. 

One thing to keep in mind is that if I try to use the value counts method again on the pandas series, for the car type column, it'll seem like nothing changed. 

In [14]:
# Notice that it looks like nothing changed
# This is because we did not update the dataframe after applying the filter
df['car_type'].value_counts()

car_type
VW Golf R         144
Toyota Sienna     120
Toyota Carolla    111
Toyota Corolla     33
Name: count, dtype: int64

And the reason why it looks like nothing changed is because we didn't assign the filtered data frame back to the original data frame. And the way to fix this is by assigning the filtered data frame back to the original data frame. 

In [15]:
# Filter dataframe to get a DataFrame of only 'Toyota Sienna'
df = df.loc[car_filter, :]

Now if you look at the value counts, it looks like we have a filtered data frame. 


In [16]:
df['car_type'].value_counts()

car_type
Toyota Sienna    120
Name: count, dtype: int64

**interest_rate Filter**

Now that we've taken care of the car type filter, we also have to make an interest rate filter. And if I look at the pandas series for the interest rate column, and the value counts for it, you'll see that we have 60 rows with a 7.02% interest rate, and 60 rows with a 3.59% interest rate. 


In [17]:
df['interest_rate'].value_counts()

interest_rate
0.0702    60
0.0359    60
Name: count, dtype: int64

And what I want to do in this section is filter the data frame to only have the 7.02% interest rate. The code here is a filter that produces a pandas series of true and false values, where the rows that are true are the ones with the 7.02% interest rate, and the false ones will be the rows with the 3.59% interest rate. So I have the data frame, I have single brackets, I have a string of the column that I'm interested in. I close those brackets. I have two equal signs, and then the 7.02% interest rate, and this produces a pandas series of true and false values. 


In [19]:
# Notice that the filter produces a pandas series of True and False values
df['interest_rate'] == 0.0702

0       True
1       True
2       True
3       True
4       True
       ...  
115    False
116    False
117    False
118    False
119    False
Name: interest_rate, Length: 120, dtype: bool

What I'm going to do next is I'm going to assign that pandas series of true and false values to the variable interest filter. To utilize my interest filter, I'm going to use the lock attribute, followed by single brackets. I'll have my pandas series of true and false values, and then I'm going to select all the columns, and I'm going to take this filtered data frame and assign it back to the original data frame. I'm going to do shift plus enter to create the filtered data frame. To check that my interest filter worked as intended, I'm going to look at the pandas series of the interest rate column. I'll use the value counts method, and I'm going to press shift plus enter, and you'll see that I have 60 rows with the 7.02% interest rate. 


In [20]:
interest_filter = df['interest_rate'] == 0.0702

In [21]:
df = df.loc[interest_filter, :]

In [22]:
df['interest_rate'].value_counts(dropna=False)

interest_rate
0.0702    60
Name: count, dtype: int64


In the previous sections, we created a car filter and an interest filter and used a lock command to filter the data by first applying the car filter and then the interest filter. A more concise way to do this is shown below. 

Bitwise Logic Operator | Meaning
--- | --- 
& | and
\| | or
^ | exclusive or
~ | not

I have my data frame, I have the lock attribute, I have single brackets, I have my car filter, and then I use the and bitwise logic operator, along with my interest filter, and then I say, I want all the columns. And this would've worked just as well as applying each of the filters individually. As you can see, by using filtering, you can get the data that you're just interested in looking at.

In [27]:
df.loc[car_filter & interest_filter].head()

Month  Starting Balance  Repayment  Interest Paid  Principal Paid  \
0      1          34689.96     687.23         202.93          484.30   
1      2          34205.66     687.23         200.10          487.13   
2      3          33718.53     687.23         197.25          489.98   
3      4          33228.55     687.23         194.38          492.85   
4      5          32735.70     687.23         191.50          495.73   

   New Balance  term  interest_rate       car_type  
0     34205.66    60         0.0702  Toyota Sienna  
1     33718.53    60         0.0702  Toyota Sienna  
2     33228.55    60         0.0702  Toyota Sienna  
3     32735.70    60         0.0702  Toyota Sienna  
4     32239.97    60         0.0702  Toyota Sienna

In [23]:
df.loc[car_filter & interest_filter, :]

Month  Starting Balance  Repayment  Interest Paid  Principal Paid  \
0       1          34689.96     687.23         202.93          484.30   
1       2          34205.66     687.23         200.10          487.13   
2       3          33718.53     687.23         197.25          489.98   
3       4          33228.55     687.23         194.38          492.85   
4       5          32735.70     687.23         191.50          495.73   
5       6          32239.97     687.23         188.60          498.63   
6       7          31741.34     687.23         185.68          501.55   
7       8          31239.79     687.23         182.75          504.48   
8       9          30735.31     687.23         179.80          507.43   
9      10          30227.88     687.23         176.83          510.40   
10     11          29717.48     687.23         173.84          513.39   
11     12          29204.09     687.23         170.84          516.39   
12     13          28687.70     687.23         167.82          519.41   
13     14          28168.29     687.23         164.78          522.45   
14     15          27645.84     687.23         161.72          525.51   
15     16          27120.33     687.23         158.65          528.58   
16     17          26591.75     687.23         155.56          531.67   
17     18          26060.08     687.23         152.45          534.78   
18     19          25525.30     687.23         149.32          537.91   
19     20          24987.39     687.23         146.17          541.06   
20     21          24446.33     687.23         143.01          544.22   
21     22          23902.11     687.23         139.82          547.41   
22     23          23354.70     687.23         136.62          550.61   
23     24          22804.09     687.23         133.40          553.83   
24     25          22250.26     687.23         130.16          557.07   
25     26          21693.19     687.23         126.90          560.33   
26     27          21132.86     687.23         123.62          563.61   
27     28          20569.25     687.23         120.33          566.90   
28     29          20002.35     687.23         117.01          570.22   
29     30          19432.13     687.23         113.67          573.56   
30     31          18858.57     687.23         110.32          576.91   
31     32          18281.66     687.23         106.94          580.29   
32     33          17701.37     687.23         103.55          583.68   
33     34          17117.69     687.23         100.13          587.10   
34     35          16530.59     687.23          96.70          590.53   
35     36          15940.06     687.23            NaN          593.99   
36     37          15346.07     687.23          89.77          597.46   
37     38          14748.61     687.23          86.27          600.96   
38     39          14147.65     687.23          82.76          604.47   
39     40          13543.18     687.23          79.22          608.01   
40     41          12935.17     687.23          75.67          611.56   
41     42          12323.61     687.23          72.09          615.14   
42     43          11708.47     687.23          68.49          618.74   
43     44          11089.73     687.23          64.87          622.36   
44     45          10467.37     687.23          61.23          626.00   
45     46           9841.37     687.23          57.57          629.66   
46     47           9211.71     687.23          53.88          633.35   
47     48           8578.36     687.23          50.18          637.05   
48     49           7941.31     687.23          46.45          640.78   
49     50           7300.53     687.23          42.70          644.53   
50     51           6656.00     687.23          38.93          648.30   
51     52           6007.70     687.23          35.14          652.09   
52     53           5355.61     687.23          31.33          655.90   
53     54           4699.71     687.23          27.49         